connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -0,0 +1,277 @@
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class DelegatingClassLoader extends URLClassLoader {
+    private static final Logger log = LoggerFactory.getLogger(DelegatingClassLoader.class);
+
+    private final ConcurrentMap<String, SortedMap<PluginDesc<?>, PluginClassLoader>> pluginLoaders;
+    private final SortedSet<PluginDesc<Connector>> connectors;
+    private final SortedSet<PluginDesc<Converter>> converters;
+    private final SortedSet<PluginDesc<Transformation>> transformations;
+    private final List<String> pluginPaths;
+    private final Map<Path, PluginClassLoader> activePaths;
+    private final Map<Path, Void> inactivePaths;
+
+    public DelegatingClassLoader(List<String> pluginPaths, ClassLoader parent) {
+        super(new URL[0], parent);
+        this.pluginPaths = pluginPaths;
+        this.pluginLoaders = new ConcurrentHashMap<>();
+        this.activePaths = new HashMap<>();
+        this.inactivePaths = new HashMap<>();
+        this.connectors = new TreeSet<>();
+        this.converters = new TreeSet<>();
+        this.transformations = new TreeSet<>();
+    }
+
+    public DelegatingClassLoader(List<String> pluginPaths) {
+        this(pluginPaths, ClassLoader.getSystemClassLoader());
+    }
+
+    private static boolean isConcrete(Class<?> cls) {
+        final int mod = cls.getModifiers();
+        return !Modifier.isAbstract(mod) && !Modifier.isInterface(mod);
+    }
+
+    private static List<Path> pluginDirs(Path topDir) throws IOException {
+        DirectoryStream.Filter<Path> dirFilter = new DirectoryStream.Filter<Path>() {
+            public boolean accept(Path path) throws IOException {
+                return Files.isDirectory(path);
+            }
+        };
+
+        List<Path> dirs = new ArrayList<>();
+        // Non-recursive for now
+        try (DirectoryStream<Path> listing = Files.newDirectoryStream(topDir, dirFilter)) {
+            for (Path dir : listing) {
+                dirs.add(dir);
+            }
+        }
+        return dirs;
+    }
+
+    private static List<URL> jarPaths(Path pluginDir) throws IOException {
+        DirectoryStream.Filter<Path> jarFilter = new DirectoryStream.Filter<Path>() {
+            public boolean accept(Path file) throws IOException {
+                return file.toString().toLowerCase(Locale.ROOT).endsWith(".jar");
+            }
+        };
+
+        List<URL> jars = new ArrayList<>();
+        try (DirectoryStream<Path> listing = Files.newDirectoryStream(pluginDir, jarFilter)) {
+            for (Path jar : listing) {
+                jars.add(jar.toUri().toURL());
+            }
+        }
+        return jars;
+    }
+
+    public void addAlias(PluginDesc<?> plugin, String alias) {
```

Comment:
I agree, this was a workaround. I'd have moved forward with adding all the alias, but didn't want to deal with name collisions at this point. I'll what's the best way to detect that the same alias exists in many loaders.